### DISCUSSION TITLE
How to set keyboard focus programatically?

### DISCUSSION THREAD

**Discussion by dov** [vsg-dev/VulkanSceneGraph#1230] [General]
Hello,

I have added the following keyboard bindings to my `trackball` to the vsgbuilder.cpp example:
```c++
    trackball->addKeyViewpoint(vsg::KeySymbol::KEY_f, lookAtDiag, 0.5);
    trackball->addKeyViewpoint(vsg::KeySymbol::KEY_7, lookAtTop, 0.5);
```
The keyboard bindings works great *after* I have pressed the mouse in the viewport once. My question is how can I get the keyboard focus set in the widget immediately after creation, i.e. without the need to focus the window by clicking in it?

**Comment by jacky-wang7** [comment]
You can try this：trackball->setViewpoint(lookAt,0.5);
This function sets the _animationDuration, and the trackball will check the animation duration at each frame.

**Reply by dov** [reply]
Thanks for the tip. However, it didn't make any difference. I still need to click once in the viewpoint before it starts responding to the key viewpoints.

**Comment by robertosfield** [comment]
This is a window manager behavior, to grap focus we'd need to add a call into each of the Window sub classes to add support for grabbing focus.  I haven't looked into how to do this with Xcb, Win32 or MacOS, if others have experience on of doing it let us know how to do it.

**Reply by dov** [reply]
I don't think it is a window manager behavior, which can be shown by adding a visitor that overrides the `vsg::KeyPressEvent()` method:

```c++
class MyKeyHandler : public vsg::Inherit<vsg::Visitor, MyKeyHandler>
{
public:
    MyKeyHandler() {}

    void apply(vsg::KeyPressEvent& keyPress) override {
        fmt::print("Key pressed:{} Modifiers:{:03x}\n",
                   (int)keyPress.keyBase,
                   (int)keyPress.keyModifier);
    }
};

viewer->addEventHandler(MyKeyHandler::create());
```

The KeyHandler reacts immediately to keyboard presses on program startup, but the trackball keyboard events react only after clicking in the view port at least once.

This was tested in X11 running under Linux.

**Reply by robertosfield** [reply]
It's a while since I working vsg::Trackball, for what I remember it's always been necessary to click on the window to give it keyboard focus.  My main development platform has been KDE and a bit of Gnome and recently Windows.

If the window has keyboard focus but the vsg::Trackball isn't yet configured to handle the events then it might be necessary to get the mouse position then use this to specify the the vsg::Trackball's vsg::Camera's ViewportState encloses the mouse position.  However, we don't presently have a get pointer position function in VSG.

Potentially if you just have a single viewport covering the whole window then you might be able to just specify the it's enabled, though as I haven't attempted it I can't provide a roadmap.
